File: translated/f-to-c/DRB046-doall2-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Two-dimensional array computation:
//Only one loop is associated with the omp for construct.
//The inner loop's loop iteration variable needs an explicit private() clause,
//otherwise it will be shared by default. No data race pairs.

#include <omp.h>
#include <stdio.h>

int main()
{
    int i, j;
    int a[100][100];

    #pragma omp parallel for private(j)
    for (i = 0; i < 100; i++) {
        for (j = 0; j < 100; j++) {
            a[i][j] = a[i][j] + 1;
        }
    }

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB046-doall2-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Two-dimensional array computation:
//Only one loop is associated with the omp for construct.
//The inner loop's loop iteration variable needs an explicit private() clause,
//otherwise it will be shared by default. No data race pairs.

package main

import (
	"runtime"
	"sync"
)

func main() {
	var a [100][100]int

	//$omp parallel do private(j)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := 100 / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= 100; start += chunkSize {
		end := start + chunkSize - 1
		if end > 100 {
			end = 100
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				for j := 1; j <= 100; j++ {
					a[i-1][j-1] = a[i-1][j-1] + 1
				}
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do
}
